vertx-web/src/main/java/io/vertx/ext/web/impl/SparseArray.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2011-2019 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+package io.vertx.ext.web.impl;
+
+/**
+ * Internal container which replaces the need for allocating
+ * expensive instances of TreeMap(s) of Integer,Handler tuples,
+ * which were being used for the sole purpose of reverse iteration
+ * on the registered handlers.
+ * This should be slightly more efficient in terms of memory
+ * pressure, including during the iteration process.
+ * Please do not reuse: this was specifically designed for a
+ * specific purpose and assumes a very dense data set.
+ *
+ * @param <H>
+ * @author Sanne Grinovero
+ */
+final class SparseArray<H> {
+
+  private Object[] elements;
+
+  void forEachInReverseOrder(final java.util.function.Consumer<H> action) {
+    if (elements != null) {
+      for (int i = elements.length - 1; i >= 0; i--) {
+        final Object element = elements[i];
+        if (element != null) {
+          action.accept((H) element);
+        }
+      }
+    }
+  }
+
+  void clear() {
+    elements = null;
```

Comment:
Clear is called explicitly when we reroute. In this case it is expected that handlers will be added again, so maybe instead of trashing the array we could just set it to nulls? And avoid future resiteToFit calls?

Reply:
Good to know! thanks, I'm missing such insight.